blocks/hooks/autocompleters.js

```diff
@@ -0,0 +1,114 @@
+/**
+ * External dependencies
+ */
+import { clone } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { addFilter, applyFilters, hasFilter } from '@wordpress/hooks';
+import { Component } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import defaultCompleters from '../autocompleters';
+
+/**
+ * Wrap the default Autocomplete component with one that
+ * supports a filter hook for customizing its list of autocompleters.
+ *
+ * Since they may be many Autocomplete instances at one time, this component
+ * applies the filter on demand, when the component is first focused after
+ * receiving a new list of completers.
+ *
+ * This function is exported directly for unit test.
+ *
+ * @param  {Function} Autocomplete Original component.
+ * @return {Function}              Wrapped component
+ */
+export function withFilteredAutocompleters( Autocomplete ) {
+	return class FilteredAutocomplete extends Component {
+		constructor() {
+			super();
+
+			this.state = {
+				completers: [],
+				hasStaleCompleters: true,
+			};
+
+			this.saveParentRef = this.saveParentRef.bind( this );
+			this.onFocus = this.onFocus.bind( this );
+		}
+
+		componentWillReceiveProps( { completers: nextCompleters } ) {
+			if ( nextCompleters !== this.props.completers ) {
+				const hasFocus = this.parentNode.contains( document.activeElement );
+
+				if ( hasFocus ) {
+					this.updateFilteredCompleters( nextCompleters );
+				} else {
+					this.expireFilteredCompleters();
+				}
+			}
+		}
+
+		updateFilteredCompleters( nextCompleters ) {
+			if ( hasFilter( 'components.Autocomplete.completers' ) ) {
+				nextCompleters = applyFilters(
+					'components.Autocomplete.completers',
+					// Provide copies so filters may directly modify them.
+					nextCompleters && nextCompleters.map( clone )
+				);
+			}
+
+			this.setState( {
+				hasStaleCompleters: false,
+				completers: nextCompleters || [],
+			} );
+		}
+
+		expireFilteredCompleters() {
+			this.setState( { hasStaleCompleters: true } );
+		}
+
+		saveParentRef( parentNode ) {
+			this.parentNode = parentNode;
+		}
+
+		onFocus() {
+			if ( this.state.hasStaleCompleters ) {
+				this.updateFilteredCompleters( this.props.completers );
+			}
+		}
+
+		render() {
+			const { completers } = this.state;
+			const autocompleteProps = {
+				...this.props,
+				completers,
+			};
+
+			return (
+				<div onFocus={ this.onFocus } ref={ this.saveParentRef }>
+					<Autocomplete onFocus={ this.onFocus } { ...autocompleteProps } />
+				</div>
+			);
+		}
+	};
+}
+
+function setDefaultCompleters( completers ) {
+	if ( ! completers ) {
+		// Provide copies so filters may directly modify them.
+		completers = defaultCompleters.map( clone );
+	}
+	return completers;
+}
+
+addFilter( 'components.Autocomplete', 'core/autocompleters', withFilteredAutocompleters );
```

Comment:
I still don't understand why we need `components.Autocomplete` filter. It is applied globally so it wraps all autocompleters. Could it be applied directly in `components/Autocomplete`? Or should we alternatively create `BlockAutocomplete` component and use it inside `RichText`?

Reply:
I don't believe there is anything here that requires the `components.Autocomplete` filter. I am trying to remember my frame of mind when I wrote this. In [this comment](https://github.com/WordPress/gutenberg/pull/4609#issuecomment-372267251), I may have thought you were suggesting we add this functionality separately via such a filter, so it is added when blocks are in use. Now that you mention it though, I suppose block and non-block UI may exist simultaneously so that we don't want to do this for all instances.

Should autocompletion extensibility be a feature limited to `@wordpress/blocks` or also supported for `@wordpress/components`?

My mind has been set on blocks, and I have not given much thought to WordPress UI outside of blocks. I don't yet know what I think about that question, but an answer will help us know what to do instead of this unnecessary filter.

If it is a blocks-only feature, I think creating a `BlockAutocomplete` component makes sense to provide the on-focus behavior currently added via the `components.Autocomplete` filter, and we should also use a blocks-related filter name instead of the generic `components.Autocomplete.completers`.

I'm thinking that the opinion of a default completers list should also be limited to blocks, but if we want autocompleters to be extensible for `@wordpress/components` as well, I am not sure how to filter and provide defaults only within a block context because the filter could be applied in both block and non-block contexts.

Reply:
I went through a similar thinking process for `MediaUpload` component in #5846. I wanted to move it to `components` module, but the feedback I received suggests to leave it inside `blocks` module together with the accompanying filer. I think it makes a lot of sense because the idea is to have a set of independent components as @youknowriad explained:
> I'm not sure we want to move components relying on filters to the components folder. I feel like the components folder is for generic React components while adding filters adds an unwanted dependency the `wp hooks`.

I would follow that advice and remove the `components.Autocomplete` filter and introduce `BlockAutocomplete` component which uses `components.Autocomplete.completers` internally.